D94/D95 instrumentation: AVRX (3 call-site hooks + per-row FFT hooks), isotropuv (2 hooks per row) and SDRAG (2 hooks around the call in DYNAM). Generated by diff -u against the PRISTINE ATMDYN.f (2026-10-05); needs ATM_DRV_dynA.f.patch (adds the ffd_avrx*/ffd_iso/ffd_sdrag helpers, units 1071-1079).

--- ATMDYN.f.orig
+++ ATMDYN.f
@@ -350,7 +350,9 @@
 #endif
       Call COMPUTE_MASS_FLUX_DIAGS (GZ, MU,MV, DT)
       call isotropuv(u,v,COS_LIMIT)
+      call ffd_sdrag(1,DTLF)
       if (USE_UNR_DRAG==0) CALL SDRAG (DTLF)
+      call ffd_sdrag(2,DTLF)
          If (MODDA < 2)  Then
             Call MAtoPMB
            CALL DIAGA
@@ -554,6 +556,7 @@
       Do 2154 J=J1XP,JNXP
       DO 2154 I=1,IM
  2154 SPA(I,J,L)=U(I,J,L)+U(I,J+1,L)
+      call ffd_avrx_site(1)
       CALL AVRX (SPA(1,J1H,L))
       I=IM
       Do 2166 J=J1XP,JNXP
@@ -1024,6 +1027,7 @@
           END DO
         END DO
 
+        call ffd_avrx_site(2)
         CALL AVRX (PU(1,J_0H,L),jrange=(/MAX(2,J_0H),MIN(JM-1,J_1H)/))
 
         DO J=J_0STG,J_1STG
@@ -1279,6 +1283,7 @@
           END DO
         END DO
 
+        call ffd_avrx_site(3)
         Call AVRX (PGFU(1,J_0H,L), JRANGE=(/Max(2,J_0H),Min(JM-1,J_1)/))
 
         DO J=J_0STG,J_1STG
@@ -1399,16 +1404,20 @@
         j0=J_0S
         j1=J_1S
       End If
+      call ffd_avrx_consts(JM,DRAT,NMIN,BYSN)
 
       DO J=j0,j1
         IF (DRAT(J).GT.1) CYCLE
+        call ffd_avrx(1,J,X(1,J),AN,BN)
         CALL FFT (X(1,J),AN,BN)
+        call ffd_avrx(2,J,X(1,J),AN,BN)
         DO N=NMIN(J),IMH-1
           AN(N)=BYSN(N)*DRAT(J) * AN(N)
           BN(N)=BYSN(N)*DRAT(J) * BN(N)
         END DO
         AN(IMH) = BYSN(IMH)*DRAT(J) * AN(IMH)
         CALL FFTI(AN,BN,X(1,J))
+        call ffd_avrx(3,J,X(1,J),AN,BN)
       END DO
 
       RETURN
@@ -1878,6 +1887,7 @@
           fac = k*dt/(dxv(j)*dxv(j))
           call shap1(ua,im,fac)
           call shap1(va,im,fac)
+          call ffd_iso(1,j,l,k,fac,u(1,j,l),v(1,j,l),ua,va)
           if(at_pole(j)) then   ! really strong filtering right at the pole
             call fft(ua,an,bn)
             an(2:imh) = 0.
@@ -1893,6 +1903,7 @@
             u(i,j,l) = cosi(i)*ua(i)+hemi*sini(i)*va(i)
             v(i,j,l) = cosi(i)*va(i)-hemi*sini(i)*ua(i)
           enddo
+          call ffd_iso(2,j,l,k,fac,u(1,j,l),v(1,j,l),ua,va)
         enddo                   ! j
       enddo                     ! l
 
